movie-service/src/main/java/com/garret/movies/service/impl/MovieServiceImpl.java

```diff
@@ -0,0 +1,135 @@
+package com.garret.movies.service.impl;
+
+import com.garret.movies.dao.entity.Movie;
+import com.garret.movies.dao.repository.MovieRepository;
+import com.garret.movies.service.api.MovieService;
+import com.garret.movies.service.exception.IncorrectDateException;
+import com.google.inject.internal.util.ImmutableList;
+import lombok.AllArgsConstructor;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+@Slf4j
+@AllArgsConstructor
+public class MovieServiceImpl implements MovieService {
+
+    private MovieRepository movieRepository;
+
+    @Override
+    @Transactional
+    public Movie save(@NonNull Movie movie) {
+        Optional<Movie> optionalMovie = movieRepository.findByImdbId(movie.getImdbId());
+        if (optionalMovie.isPresent()) {
+            log.info(movie.getTitle() + " already present");
+            return optionalMovie.get();
+        }
+        log.info("Start saving movie " + movie.getTitle());
+        Movie savedMovie = movieRepository.save(movie);
+        log.info(movie.getTitle() + " was saved to database");
+        return savedMovie;
+    }
+
+    @Override
+    @Transactional
+    public List<Movie> saveAll(@NonNull List<Movie> movies) {
+        movieRepository.saveAll(movies);
+        log.info(movies.size() + " movies saved to database");
+        return movies;
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public Optional<Movie> getById(@NonNull Long id) {
+        return movieRepository.findById(id);
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<Movie> getAll() {
+        return ImmutableList.copyOf(movieRepository.findAll());
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public Optional<Movie> getByImdbId(@NonNull String imdbId) {
+        return movieRepository.findByImdbId(imdbId);
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<Movie> getAllByGenre(@NonNull String genre) {
+        return movieRepository.findAllByGenresValue(genre);
+    }
+
+    @Override
+    public List<Movie> getAllByActor(@NonNull String actor) {
+        return movieRepository.findAllByActorsFullNameContains(actor);
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<Movie> getAllByYear(@NonNull Integer year) {
+        String start = year + "-01-01";
+        String end = year + "-12-31";
+        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd");
```

Comment:
@Garret-vin If you go inside SimpleDateFormat you will see that during initialization it creates regex and initialize it which takes time and resources. As you have static format template its better to move it in final class variable and init only once instead of doing that each time